builder/virtualbox: Error removing iso due to virtualbox lock still being hold

#### Overview of the Issue

Packer is having errors removing/ejecting the guest additions iso from the device as you can see at the logs bellow.

This issue seems related to https://github.com/hashicorp/packer/issues/4432, which seems to hint that the problem also applies/happens to cdroms.

#### Reproduction Steps

packer build

### Packer version

1.4.4

### Simplified Packer Buildfile

```
...
    {
      "name": "ubuntu-{{user `version`}}-amd64-virtualbox",
      "type": "virtualbox-iso",
      "guest_os_type": "Ubuntu_64",
      "guest_additions_mode": "attach",
      "headless": true,
...
```

The full file is at https://github.com/rgl/ubuntu-vagrant/blob/master/ubuntu.json#L49.

### Operating system and Environment details

Packer is running on Windows 10 and launching an Ubuntu 18.04 guest in VirtualBox 6.0.14.

### Log Fragments and crash.log files

I've started having these errors removing the guest additions iso:

```
2019/10/27 09:16:08 packer.exe: 2019/10/27 09:16:08 [ERROR] Remote command exited without exit status or exit signal.
2019/10/27 09:16:08 ui error: [1;31m==> ubuntu-18.04-amd64-virtualbox: [sudo] password for vagrant:[0m
2019/10/27 09:16:08 packer.exe: 2019/10/27 09:16:08 Waiting max 5m0s for shutdown to complete
2019/10/27 09:16:09 packer.exe: 2019/10/27 09:16:09 VM shut down.
2019/10/27 09:16:09 packer.exe: 2019/10/27 09:16:09 Executing VBoxManage: []string{"storageattach", "packer-ubuntu-18.04-amd64-virtualbox-1572167140", "--storagectl", "IDE Controller", "--port", "0", "--device", "1", "--medium", "none"}
2019/10/27 09:16:09 packer.exe: 2019/10/27 09:16:09 stdout:
2019/10/27 09:16:09 packer.exe: 2019/10/27 09:16:09 stderr: VBoxManage.exe: error: Failed to get a console object from the direct session (VBOX_E_INVALID_OBJECT_STATE)
2019/10/27 09:16:09 packer.exe: VBoxManage.exe: error: Details: code VBOX_E_VM_ERROR (0x80bb0003), component MachineWrap, interface IMachine, callee IUnknown
2019/10/27 09:16:09 packer.exe: VBoxManage.exe: error: Context: "LockMachine(a->session, LockType_Shared)" at line 336 of file VBoxManageStorageController.cpp
2019/10/27 09:16:09 ui error: [1;31m==> ubuntu-18.04-amd64-virtualbox: Error detaching ISO: VBoxManage error: VBoxManage.exe: error: Failed to get a console object from the direct session (VBOX_E_INVALID_OBJECT_STATE)
==> ubuntu-18.04-amd64-virtualbox: VBoxManage.exe: error: Details: code VBOX_E_VM_ERROR (0x80bb0003), component MachineWrap, interface IMachine, callee IUnknown
==> ubuntu-18.04-amd64-virtualbox: VBoxManage.exe: error: Context: "LockMachine(a->session, LockType_Shared)" at line 336 of file VBoxManageStorageController.cpp[0m
2019/10/27 09:16:09 ui error: [1;31m==> ubuntu-18.04-amd64-virtualbox: Error detaching ISO: VBoxManage error: VBoxManage.exe: error: Failed to get a console object from the direct session (VBOX_E_INVALID_OBJECT_STATE)
==> ubuntu-18.04-amd64-virtualbox: VBoxManage.exe: error: Details: code VBOX_E_VM_ERROR (0x80bb0003), component MachineWrap, interface IMachine, callee IUnknown
==> ubuntu-18.04-amd64-virtualbox: VBoxManage.exe: error: Context: "LockMachine(a->session, LockType_Shared)" at line 336 of file VBoxManageStorageController.cpp[0m
2019/10/27 09:16:09 ui error: [1;31m==> ubuntu-18.04-amd64-virtualbox: Step "StepRemoveDevices" failed, aborting...[0m
```

Setting `"post_shutdown_delay": "2m"` works around the problem, but doing that feels more like an workaround rather than a root cause fix.

The problem might even be on VirtualBox (sorry, but I didn't try with an older version).

--- a/builder/virtualbox/common/driver_4_2.go
+++ b/builder/virtualbox/common/driver_4_2.go
@@ -203,11 +203,11 @@ func (d *VBox42Driver) VBoxManageWithOutput(args ...string) (string, error) {
 
 	ctx := context.TODO()
 	err := retry.Config{
-		Tries: 11,
+		Tries: 5,
 		ShouldRetry: func(err error) bool {
-			return isSessionLocked(stderr)
+			return isLockedBySession(stderr)
 		},
-		RetryDelay: (&retry.Backoff{InitialBackoff: 30 * time.Second, MaxBackoff: 2 * time.Minute, Multiplier: 2}).Linear,
+		RetryDelay: (&retry.Backoff{InitialBackoff: 1 * time.Minute, MaxBackoff: 2 * time.Minute, Multiplier: 2}).Linear,
 	}.Run(ctx, func(ctx context.Context) error {
 		return cmd.Run()
 	})
@@ -234,9 +234,9 @@ func (d *VBox42Driver) VBoxManageWithOutput(args ...string) (string, error) {
 	return stdoutString, err
 }
 
-func isSessionLocked(stderr bytes.Buffer) bool {
+func isLockedBySession(stderr bytes.Buffer) bool {
 	stderrString := strings.TrimSpace(stderr.String())
-	return strings.Contains(stderrString, "VBOX_E_INVALID_OBJECT_STATE") && strings.Contains(stderrString, "LockMachine")
+	return strings.Contains(stderrString, "LockMachine") || strings.Contains(stderrString, "already locked by a session")
 }
 
 func (d *VBox42Driver) Verify() error {

--- a/builder/virtualbox/common/driver_4_2.go
+++ b/builder/virtualbox/common/driver_4_2.go
@@ -200,7 +200,17 @@ func (d *VBox42Driver) VBoxManageWithOutput(args ...string) (string, error) {
 	cmd := exec.Command(d.VBoxManagePath, args...)
 	cmd.Stdout = &stdout
 	cmd.Stderr = &stderr
-	err := cmd.Run()
+
+	ctx := context.TODO()
+	err := retry.Config{
+		Tries: 11,
+		ShouldRetry: func(err error) bool {
+			return isSessionLocked(stderr)
+		},
+		RetryDelay: (&retry.Backoff{InitialBackoff: 30 * time.Second, MaxBackoff: 2 * time.Minute, Multiplier: 2}).Linear,
+	}.Run(ctx, func(ctx context.Context) error {
+		return cmd.Run()
+	})
 
 	stdoutString := strings.TrimSpace(stdout.String())
 	stderrString := strings.TrimSpace(stderr.String())
@@ -222,6 +232,11 @@ func (d *VBox42Driver) VBoxManageWithOutput(args ...string) (string, error) {
 	log.Printf("stderr: %s", stderrString)
 
 	return stdoutString, err
+}
+
+func isSessionLocked(stderr bytes.Buffer) bool {
+	stderrString := strings.TrimSpace(stderr.String())
+	return strings.Contains(stderrString, "VBOX_E_INVALID_OBJECT_STATE") && strings.Contains(stderrString, "LockMachine")
 }
 
 func (d *VBox42Driver) Verify() error {

--- a/builder/virtualbox/common/driver_4_2.go
+++ b/builder/virtualbox/common/driver_4_2.go
@@ -165,9 +165,6 @@ func (d *VBox42Driver) Stop(name string) error {
 		return err
 	}
 
-	// We sleep here for a little bit to let the session "unlock"
-	time.Sleep(2 * time.Second)
-
 	return nil
 }
 
@@ -207,7 +204,7 @@ func (d *VBox42Driver) VBoxManageWithOutput(args ...string) (string, error) {
 		ShouldRetry: func(err error) bool {
 			return isLockedBySession(stderr)
 		},
-		RetryDelay: (&retry.Backoff{InitialBackoff: 1 * time.Minute, MaxBackoff: 2 * time.Minute, Multiplier: 2}).Linear,
+		RetryDelay: func() time.Duration { return 2 * time.Minute },
 	}.Run(ctx, func(ctx context.Context) error {
 		return cmd.Run()
 	})
@@ -236,7 +233,7 @@ func (d *VBox42Driver) VBoxManageWithOutput(args ...string) (string, error) {
 
 func isLockedBySession(stderr bytes.Buffer) bool {
 	stderrString := strings.TrimSpace(stderr.String())
-	return strings.Contains(stderrString, "LockMachine") || strings.Contains(stderrString, "already locked by a session")
+	return strings.Contains(stderrString, "VBOX_E_INVALID_OBJECT_STATE")
 }
 
 func (d *VBox42Driver) Verify() error {

--- a/builder/virtualbox/common/driver_4_2.go
+++ b/builder/virtualbox/common/driver_4_2.go
@@ -190,7 +190,7 @@ func (d *VBox42Driver) VBoxManage(args ...string) error {
 	err := retry.Config{
 		Tries: 5,
 		ShouldRetry: func(err error) bool {
-			return isLockedBySession(err.Error())
+			return strings.Contains(err.Error(), "VBOX_E_INVALID_OBJECT_STATE")
 		},
 		RetryDelay: func() time.Duration { return 2 * time.Minute },
 	}.Run(ctx, func(ctx context.Context) error {
@@ -230,10 +230,6 @@ func (d *VBox42Driver) VBoxManageWithOutput(args ...string) (string, error) {
 	log.Printf("stderr: %s", stderrString)
 
 	return stdoutString, err
-}
-
-func isLockedBySession(err string) bool {
-	return strings.Contains(err, "VBOX_E_INVALID_OBJECT_STATE")
 }
 
 func (d *VBox42Driver) Verify() error {

--- a/builder/virtualbox/common/driver_4_2.go
+++ b/builder/virtualbox/common/driver_4_2.go
@@ -186,7 +186,18 @@ func (d *VBox42Driver) SuppressMessages() error {
 }
 
 func (d *VBox42Driver) VBoxManage(args ...string) error {
-	_, err := d.VBoxManageWithOutput(args...)
+	ctx := context.TODO()
+	err := retry.Config{
+		Tries: 5,
+		ShouldRetry: func(err error) bool {
+			return isLockedBySession(err.Error())
+		},
+		RetryDelay: func() time.Duration { return 2 * time.Minute },
+	}.Run(ctx, func(ctx context.Context) error {
+		_, err := d.VBoxManageWithOutput(args...)
+		return err
+	})
+
 	return err
 }
 
@@ -197,17 +208,7 @@ func (d *VBox42Driver) VBoxManageWithOutput(args ...string) (string, error) {
 	cmd := exec.Command(d.VBoxManagePath, args...)
 	cmd.Stdout = &stdout
 	cmd.Stderr = &stderr
-
-	ctx := context.TODO()
-	err := retry.Config{
-		Tries: 5,
-		ShouldRetry: func(err error) bool {
-			return isLockedBySession(stderr)
-		},
-		RetryDelay: func() time.Duration { return 2 * time.Minute },
-	}.Run(ctx, func(ctx context.Context) error {
-		return cmd.Run()
-	})
+	err := cmd.Run()
 
 	stdoutString := strings.TrimSpace(stdout.String())
 	stderrString := strings.TrimSpace(stderr.String())
@@ -231,9 +232,8 @@ func (d *VBox42Driver) VBoxManageWithOutput(args ...string) (string, error) {
 	return stdoutString, err
 }
 
-func isLockedBySession(stderr bytes.Buffer) bool {
-	stderrString := strings.TrimSpace(stderr.String())
-	return strings.Contains(stderrString, "VBOX_E_INVALID_OBJECT_STATE")
+func isLockedBySession(err string) bool {
+	return strings.Contains(err, "VBOX_E_INVALID_OBJECT_STATE")
 }
 
 func (d *VBox42Driver) Verify() error {

--- a/builder/virtualbox/common/shutdown_config.go
+++ b/builder/virtualbox/common/shutdown_config.go
@@ -41,5 +41,9 @@ func (c *ShutdownConfig) Prepare(ctx *interpolate.Context) []error {
 		c.ShutdownTimeout = 5 * time.Minute
 	}
 
+	if c.PostShutdownDelay == 0 {
+		c.PostShutdownDelay = 2 * time.Second
+	}
+
 	return nil
 }

--- a/builder/virtualbox/common/shutdown_config_test.go
+++ b/builder/virtualbox/common/shutdown_config_test.go
@@ -50,7 +50,7 @@ func TestShutdownConfigPrepare_PostShutdownDelay(t *testing.T) {
 		t.Fatalf("err: %#v", errs)
 	}
 	if c.PostShutdownDelay != 2*time.Second {
-		t.Fatalf("bad PostShutdownDelay should be 2 seconds but was %s", c.PostShutdownDelay)
+		t.Fatalf("bad: PostShutdownDelay should be 2 seconds but was %s", c.PostShutdownDelay)
 	}
 
 	// Test with a good one

--- a/builder/virtualbox/common/shutdown_config_test.go
+++ b/builder/virtualbox/common/shutdown_config_test.go
@@ -49,8 +49,8 @@ func TestShutdownConfigPrepare_PostShutdownDelay(t *testing.T) {
 	if len(errs) > 0 {
 		t.Fatalf("err: %#v", errs)
 	}
-	if c.PostShutdownDelay.Nanoseconds() != 0 {
-		t.Fatalf("bad: %s", c.PostShutdownDelay)
+	if c.PostShutdownDelay != 2*time.Second {
+		t.Fatalf("bad PostShutdownDelay should be 2 seconds but was %s", c.PostShutdownDelay)
 	}
 
 	// Test with a good one

